### ISSUE TITLE
dnSpy (Roslyn) crashes (Assertion Failed) when trying to get local name of compiler generated variable:  static array in VB

### ISSUE THREAD

**Issue by ecniVcinimoD** [0xd4d/dnSpy#1147]
dnSpy is throwing an exception in the following class:

namespace dnSpy.Roslyn.Debugger.ExpressionCompiler {
	static class RoslynExpressionCompilerMethods {

when trying to get the local name of a compiler generated variable.  The failing line is here:
			foreach (var info in compilerGeneratedVariables) {
				Debug.Assert(res[info.Index] == null);
in the function:
		public static ImmutableArray<string> GetLocalNames(int totalLocals, List<DbgMethodDebugScope> scopes, CompilerGeneratedVariableInfo[] compilerGeneratedVariables) {

It appears that the element at the identified index is already filled by the code above this which looks at scopes[0].Locals.Index.

What type of logical mistake can cause filling the same index twice?
Do we need other "continue" statements in the scopes[0].Locals.Index loop?

**Comment by ecniVcinimoD** [comment]
Update:  this failure is associated with Static variable declarations.  If I change the Static declarations to Dim then the execution failure doesn't happen.  Of course my code doesn't run as expected but at least dnSpy does not crash.

**Comment by ecniVcinimoD** [comment]
It seems that this bug can be fixed with the following change to the GetLocalNames routine such that the OrElse logical check to see if the local IsDebuggerHidden is added:
					if (local.IsDecompilerGenerated  || local.IsDebuggerHidden)

Below works for me:


		public static ImmutableArray<string> GetLocalNames(int totalLocals, List<DbgMethodDebugScope> scopes, CompilerGeneratedVariableInfo[] compilerGeneratedVariables) {
			if (totalLocals == 0)
				return ImmutableArray<string>.Empty;
			var res = new string[totalLocals];
			foreach (var scope in scopes) {
				foreach (var local in scope.Locals) {
					if (local.IsDecompilerGenerated  || local.IsDebuggerHidden)
						continue;
					if (local.Index < 0)
						continue;

					res[local.Index] = local.Name;
				}
			}
			foreach (var info in compilerGeneratedVariables) {
				Debug.Assert(res[info.Index] == null);
				if (res[info.Index] == null)
					res[info.Index] = info.Name;
			}
			return ImmutableArray.Create(res);
		}

**Comment by ecniVcinimoD** [comment]
Is there a way to show less decoration in Static variables?  I see the following:
Me.$STATIC$OuterLoop$2061C12211225D822$dQdOSNRDict.Add(48, 0.456)

But would rather see:
dQdOSNRDict.Add(48, 0.456)

Is there a setting that allows me to clean up the Static appearance in the debug window?

**Comment by 0xd4d** [comment]
Do you have a repro for this issue?

**Comment by ecniVcinimoD** [comment]
The dynamically defined code that I'm debugging has STATIC declarations (in VB.NET).  I believe that is all that is required to cause the bug to appear.  The fix I suggested above seems to have solved the problem for me.

**Comment by 0xd4d** [comment]
I mean, do you have simple VB code you could paste here that I can compile and debug to repro this.

**Comment by ecniVcinimoD** [comment]
Well this is bizarre.  I had a recipe for recreating the problem every time and now I cannot recreate the bug.  I created an installer (so I could deploy the executable from a shared network directory) and I changed the XmlSettingsReader.cs to ignore previous temporary DLLs from my scripting application (I recompile the test code every time the user runs their code; I open the most recent DLL using the command line to open dnSpy.).

Something changed on my computer such that the bug doesn't show now.

Sorry, I wish I understood why things are different now than they were on the very frustrating day when I submitted the issue.
